### ISSUE TITLE
Armor Penetration is stronger than intended

### ISSUE THREAD

**Issue by Undertower** [dalaranwow/dalaran-wow#5646]
[//]: # (INFORMATION ABOUT TAGS)
[//]: # (The black and yellow tag is unique. There can only be one label assigned to them - one black and one yellow)
[//]: # (Priority - ... tags are assigned by team members)
[//]: # (Your issue will be reviewed by testers and the developers will fix it in the future.)
[//]: # (Only once the issue got the label [Fixed - On Live Server] you can find the fix on your Gameserver.)
[//]: # (You can increase the relevance of your issue by adding a thumb-up emote http://i.imgur.com/RFouVVi.png)

[//]: # (***************************************************************)
[//]: # (** !!!!!!!!!!!!!DON'T DELETE THIS TEMPLATE ELSE YOUR ISSUE WILL BE CLOSED!!!!!!!!!!!! **)
[//]: # (***************************************************************)

**Current behaviour**: At 1400 ArP rating the damage reduction from Boss Armor is reducing the damage by 9.343%. When using Expose Armor and Faerie Fire the damage reduction from Boss Armor is 0%. 

**Expected behaviour**: At 1400 ArP rating the damage reduction from Boss Armor should reduce the damage of physical abilities by 11.70%. Using Faerie Fire and Expose Armor/Sunder Armor the damage reduction should be at 2.03%

**Steps to reproduce the problem**:

1. Get 1400 ArP.
2. Get a melee weapon that has a small damage range(a 1-2 damage melee weapon) or a ranged weapon that has the same min-max damage([Fine Light Crossbow](https://wotlk.evowow.com/?item=15808))
3. Compare the Actual damage to the Expected damage.

**Include proofs for this behaviour**

Auto Shot damage = Weapon_Damage + (RAP * Weapon_Speed) + (Ammo_Damage * Weapon_Speed) * [1.05](https://wotlk.evowow.com/?spell=19509)

1489,257 = (29 + 1142,29 + 247.05) * [1.05](https://wotlk.evowow.com/?spell=19509)

[Here](https://i.imgur.com/DXfA0Ee.jpg) you can see the Auto Shot damage in my character sheet and the Auto Shot damage on the target.

At 1400 ArP the damage is expected to be reduced by 11.70%
1489 * 0.883 = 1314,787

However the damage is reduced 9.343%:
1489 * 0.90657 = 1349.88273

The ArP formula without debuffs on retail is as follows:

Max Armor reduced by ArP: (10643 + 15232.5) / 3 = 8625
Current Armor reduction by ArP: 10643 - (8625 * 100%) = 2018
Damage reduction from Boss Armor: 2018 / (2018 + 15232.5) = 11.698%

ArP formula without debuffs on Dalaran-wow is as follows:

Max Armor reduced by ArP: ([10673](https://i.imgur.com/qGLlSh9.jpg) + 16635) / 3 = 9102.66
Current Armor reduction by ArP: [10673](https://i.imgur.com/qGLlSh9.jpg) - (9102.66 * 100%) = 1570.34
Damage reduction from Boss Armor: 1570.34 / (1570.34 + 15232.5) = 9.343%

[Here](https://i.imgur.com/pC0rbvp.jpg) is a screenshot of my damage on a target with Expose Armor and Faerie fire. 

ArP formula from retail with debuffs:
Armor reduction from Debuffs: 10643 * (1 - [0.05](https://wotlk.evowow.com/?spell=770)) * (1 - [0.2](https://wotlk.evowow.com/?spell=8647)) = 8088.68
Max Armor reduced by ArP with debuffs:  (8088.68 + 15232.5) / 3 = 7773.73
Current Armor reduction by ArP: 8088.68 - (7773.73 * 100%) = 314.95
Damage reduction from Boss Armor: 315 / (315 + 15232.5) = 2.03%

Actual ArP formula with debuffs on Dalaran-wow: 
Armor reduction from Debuffs: [10673](https://i.imgur.com/qGLlSh9.jpg) * (1 - [0.05](https://wotlk.evowow.com/?spell=770)) * (1 - [0.2](https://wotlk.evowow.com/?spell=8647)) = 8111.48
Max Armor reduced by ArP with debuffs: (8111.48 + 16635) / 3 = 8248.82
Current Armor reduction by ArP: 8111.48 - (8111.48 (Should be 8248.82 but Total Armor Penetration from ArP Rating can not be higher than Debuffed Armor) * 100% ) = 0

**Comment by Chenocide** [comment]
The problem is most likely related to this comment on TrinityCore:
https://github.com/TrinityCore/TrinityCore/issues/15024

This means that the armor constant is inverted and the armor constant for level 80 attackers is instead applied to level 83 attackers and vice versa. The code example below from TrinityCore shows that the armor constant is being calculated with the victims level in mind rather than the attacker's level. 

            `float maxArmorPen = 0.f;
            if (victim->getLevel() < 60)
                maxArmorPen = float(400 + 85 * victim->getLevel());
            else
                maxArmorPen = 400 + 85 * victim->getLevel() + 4.5f * 85 * (victim->getLevel() - 59);
            
            // Cap armor penetration to this number
            maxArmorPen = std::min((armor + maxArmorPen) / 3.f, armor);
            // Figure out how much armor do we ignore
            armor -= CalculatePct(maxArmorPen, arpPct);`


This means that there's a slight gain to damage done by armor penetration. 
Full calculation and explanation below:
I'm using a 1 handed weapon on an Enhancement shaman with 1.9 speed and 4349 attack power and 10% damage done from talents. 

Here's the damage calculations for this enhancement.
![](https://puu.sh/CgswJ/fb2e2a4571.png)

Theoretical weapon damage - Almost 100% identical to character weapon damage
![](https://puu.sh/CgsAh/2e6ecc7e30.png)

First test with no armor penetration. 
![](https://puu.sh/CgsCz/c4437a3fda.png)

Applicable armor is calculated by as seen from the code example above. The problem is caused by using that armor constant of the target, instead of the attacker. In a raid scenario, the target would be level 83 and the constant is therefore 16635, instead of 15232.5, 
(10643 + 16635) / 3 = 9092.66

Since there's no armor penetration is this exact test, we can just say:

650.342/100 * 41.13 = 267.52
and
650.342 - 267.52 = 382.9. 

First test with 11.07% armor penetration.
![](https://puu.sh/CgsP0/ceb3064d61.png)

Mostly the same stuff but with a few extra steps. We need to calculate the effective armor.

EA: 10110.85 - (9092.66 * 11.07%) = 9636.44

And finally the damage reduction:

DR: 9636.44 / (9636.44 + 15232.5) = 0.3874 or basically 38.75% damage reduction. 
The armor constant of a level 83 boss (16635) is not used here. 

650.342/100 * 38.75 = 267.52
and
650.342 - 267.52 = 398.33

The damage values matches up almost perfectly with the math. 
On a side note, there's a copy of TrinityCore on Github that I expect to be near 100% identical to Dalaran-Wow and it's the exact same problem there. 

TL:DR:

The armor constant is inverted and the armor constant for level 80 attackers is instead applied to level 83 attackers and vice versa. The code example below from TrinityCore shows that the armor constant is being calculated with the victims level in mind rather than the attacker's level.

**Comment by Chenocide** [comment]
Dalaran-wow most likely uses the wrong armor value on all bosses:

https://github.com/TrinityCore/TrinityCore/issues/15106

If I use the correct armor value in my calculations, then the calculated damage comes even closer to the observed damage.

**Comment by Lothloryen** [comment]
Hello there,

Thank you very much for your so elaborated report. I have been able to deploy some pending fixes that affects this matter now, it should be fine.

Regards.

**Comment by Chenocide** [comment]
Armor penetration seems to still have issues. 
Here's the calculations for a druid with 45.37% arp, 2512 AP. 
![](https://i.imgur.com/KAWZWID.png)

In-game dummy tests shows that I'm doing more damage than I should with the lvl 80 armor constant.
Max damage with normal white hit: 
![](https://i.imgur.com/qUktSKR.png)

It would be explained by the level 83 constant, 16635.
@Lothloryen

**Comment by Undertower** [comment]
I have come to the same conclusion. 
Using the retail formula, the damage at 100% Armor Penetration does not match up.

Retail Armor Penetration formula:
Armor Penetration Cap: (10643 + 15235.5) / 3 = 8625.166
Effective Armor: 10643 - (8625.166 * 100%) = 2017.834
Damage Reduction: 2017.834 / (2017.834 + 15232.5) = 11.697%

If you look at my screenshot attached below, you can see that the damage of my character sheet is 1603-1604.
At 100% ArP with no debuffs, the Boss armor should reduce the damage by 11.697%.

1603 * 0.88303 = 1415.497
1604 * 0.88303 = 1416.380

As you can see, neither of those numbers match up to the damage I have recorded on my recount.

The issue is the same as previously, except now the armor value of bosses is correct.

Here's the current, faulty, formula on dalaran-wow:
Armor Penetration Cap: (10643 + 16635) / 3 = 9092.666
Effective Armor: 10643 - (9092.66 * 100%) = 1550.334
Damage Reduction: 1550.334 / (1550.334 + 15232.5) = 9.237%

1603 * 0.90763 = 1454.930
1604 * 0.90763 =  1455.838
![WoWScrnShot_060319_034053](https://user-images.githubusercontent.com/44290478/58771334-d757b380-85b3-11e9-9d27-efe3143068dc.jpg)
